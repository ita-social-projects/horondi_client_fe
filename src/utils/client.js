import { ApolloClient, gql, split, ApolloLink, InMemoryCache } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { setContext } from '@apollo/client/link/context';
import fetch from 'unfetch';
import { IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import { createUploadLink } from 'apollo-upload-client/public';
import { getMainDefinition } from '@apollo/client/utilities';

import { AUTH_ERRORS, USER_TOKENS } from '../configs';
import { getFromLocalStorage, setToLocalStorage } from '../services/local-storage.service';
import refreshAuthToken from './regenerateAuthTokenPair';

const introspectionResult = require('../fragmentTypes');

const { ACCESS_TOKEN } = USER_TOKENS;
const { ACCESS_TOKEN_IS_NOT_VALID } = AUTH_ERRORS;
const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: introspectionResult
});

export const REACT_APP_API_URL =
  window.env && window.env.REACT_APP_API_URL
    ? window.env.REACT_APP_API_URL
    : process.env.REACT_APP_API_URL;

const wsLink = new WebSocketLink({
  uri: process.env.REACT_APP_WS_API_URL,
  options: {
    reconnect: true,
    connectionParams: {
      authToken: getFromLocalStorage(ACCESS_TOKEN)
    }
  }
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    token: getFromLocalStorage(ACCESS_TOKEN) || ''
  }
}));

const terminatingLink = split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'subscription';
  },
  wsLink,
  authLink.concat(createUploadLink({ uri: `${REACT_APP_API_URL}/graphql` }))
);

export const client = new ApolloClient({
  fetch,
  link: ApolloLink.from([terminatingLink]),
  cache: new InMemoryCache({
    addTypename: false,
    fragmentMatcher
  })
});

export const getItems = async (query, variables = {}) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const queryResult = await client.query({
      query: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (queryResult.data && Object.values(queryResult.data)[0]?.message) {
      const message = Object.values(queryResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await getItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return queryResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setItems = async (query, variables) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const mutationResult = await client.mutate({
      mutation: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (mutationResult.data && Object.values(mutationResult.data)[0]?.message) {
      const message = Object.values(mutationResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await setItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return mutationResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setChatMail = (query, variables) =>
  client.mutate({
    mutation: gql`
      ${query}
    `,
    variables
  });
