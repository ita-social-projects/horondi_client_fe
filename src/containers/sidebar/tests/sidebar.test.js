import React from 'react';
import { useQuery } from '@apollo/client';

import Sidebar from '../sidebar';

let wrapper;
const useQueryData = {
  loading: false,
  error: false,
  data: { getCategoriesForBurgerMenu: [{ category: { _id: 1 } }] }
};
const props = { fromSideBar: {}, isMenuOpen: true, setIsMenuOpen: () => null };

jest.mock('@apollo/client');
jest.mock('../../sidemenu-right-bar', () => ({
  __esModule: true,
  default: () => null
}));

jest.mock('@material-ui/styles', () => ({
  ...jest.requireActual('@material-ui/styles'),
  useTheme: () => ({
    palette: {
      type: 'light',
      cart: {
        borderColor: '#000000'
      }
    }
  })
}));

describe('sidebar tests', () => {
  it('should be defined', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData
    }));
    wrapper = shallow(<Sidebar {...props} />);

    expect(wrapper).toBeDefined();
  });

  it('should match snapshot', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData
    }));

    wrapper = shallow(<Sidebar {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should cover other branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      loading: true
    }));

    wrapper = shallow(<Sidebar {...props} />);
  });

  it('should cover rest branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      error: {}
    }));

    wrapper = shallow(<Sidebar {...props} />);
  });
});
