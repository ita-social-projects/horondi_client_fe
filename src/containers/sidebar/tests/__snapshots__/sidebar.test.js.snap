// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar tests should match snapshot 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  className="makeStyles-drawer-1 makeStyles-drawer-10"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(List)) />
  <Link
    className="makeStyles-mainItem-3"
    onClick={[Function]}
    to="/constructor"
  >
    <span
      className="makeStyles-constructorItem-7"
    >
      sidebar.constructorCreate
    </span>
  </Link>
  <div
    className="makeStyles-subList-4"
  >
    <Link
      className="makeStyles-subItem-5"
      key="/materials"
      onClick={[Function]}
      to="/materials"
    >
      <span>
        Materials
      </span>
    </Link>
    <Link
      className="makeStyles-subItem-5"
      key="/pages/about-us"
      onClick={[Function]}
      to="/pages/about-us"
    >
      <span>
        About us
      </span>
    </Link>
    <Link
      className="makeStyles-subItem-5"
      key="/news"
      onClick={[Function]}
      to="/news"
    >
      <span>
        News
      </span>
    </Link>
    <Link
      className="makeStyles-subItem-5"
      key="/contacts"
      onClick={[Function]}
      to="/contacts"
    >
      <span>
        Contacts
      </span>
    </Link>
  </div>
  <SocialLinks
    position="center"
    setIsMenuOpen={[Function]}
    socialIconsStyles="makeStyles-socialIconsStyles-6"
  />
  <default
    fromSideBar={true}
    setIsMenuOpen={[Function]}
  />
</WithStyles(ForwardRef(Drawer))>
`;
